### PR TITLE
Merge ASYNC-branch into MAIN

### PR DESCRIPTION
**Pull request by bouwew**
## Summary by CodeRabbit

- **New Features**
  - Complete asynchronous redesign with modern type-safe API for Plugwise USB-Stick integration.
  - Added comprehensive local caching for nodes and network registrations.
  - Enhanced node support with async load, initialize, and state management for Circle, Circle Plus, Sense, Scan, Switch, Celsius, and SED devices.
  - New subscription mechanism for node feature updates and event-driven architecture.
  - Detailed energy pulse and calibration helpers for improved energy data handling.
  - Added extensive enums, data classes, and protocols defining device features and states.
  - Introduced async connection manager, sender, receiver, and prioritized request queue for USB-Stick communication.
  - Added helper modules for utilities, caching, firmware support, and subscription management.

- **Bug Fixes**
  - Fixed edge cases in device discovery, cache handling, and energy log management.
  - Corrected minor typos and improved error handling across modules.

- **Refactor**
  - Removed synchronous, thread-based connection and message controller in favor of async connection components.
  - Replaced legacy parser with async message processing pipeline.
  - Streamlined message request and response classes with explicit async send and await patterns.
  - Overhauled node implementations to async classes with improved caching and event subscriptions.
  - Simplified constants and error classes for clarity and maintainability.
  - Consolidated node factory for type-based instantiation.

- **Chores**
  - Updated dependencies to use pyserial-asyncio-fast and aiofiles; dropped deprecated packages.
  - Adjusted Python version support to 3.13.
  - Cleaned up project metadata and configuration files.
  - Added helper modules for utilities, caching, and firmware support.
  - Updated GitHub workflows and pre-commit configurations.

- **Documentation**
  - Refined changelog entries detailing major async rewrite and feature additions.
  - Added comprehensive docstrings and type annotations throughout.

- **Tests**
  - Introduced extensive binary test data for USB stick message-response validation.

- **Revert**
  - Removed legacy synchronous connection, parser, and controller modules.

### CODE REVIEW

**Comment by CoMPaTech** [comment]
🥳 #itshappening

File: plugwise_usb/nodes/helpers/pulses.py [hunk 1]
```diff
@@ -0,0 +1,1036 @@
...
+                self._rollover_production = True
+
+        self._pulses_consumption = pulses_consumed
+        self._pulses_production = pulses_produced
+        _LOGGER.debug(
+            "update_pulse_counter | consumption pulses=%s | production pulses=%s",
+            self._pulses_consumption,
+            self._pulses_production,
+        )
+
```
[the remark this answers is not in the document]
Comment (bouwew) [reply plugwise_usb/nodes/helpers/pulses.py:313]: `self_pulses_production` and `pulses_produced` are negative values while `self._pulses_consumption` and `pulses_consumed` are positive values. So the comparison operator is correct.

File: plugwise_usb/nodes/helpers/pulses.py [hunk 2]
```diff
@@ -0,0 +1,1036 @@
...
+        last_known_timestamp = self._logs[address][slot].timestamp
+        address, slot = calc_log_address(address, slot, -1)
+        while (
+            self._log_exists(address, slot)
+            and self._logs[address][slot].timestamp == last_known_timestamp
+        ):
+            address, slot = calc_log_address(address, slot, -1)
+
+        return self._logs[address][slot].timestamp - last_known_timestamp
+
```
[the remark this answers is not in the document]
Comment (bouwew) [reply plugwise_usb/nodes/helpers/pulses.py]: Nice find!

**Comment by ArnoutD** [comment]
you are on (fire) probeer je te volgen maar de branches vliegen me om de oren

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,18 +1,42 @@
...
+
+Full rewrite of library into async version (@brefra).
+Main list of changes:
+- Full async and typed
+- Improved protocol handling
+- Support for local caching of collected data to improve startup and device detection
+- Improved handling of edge cases especially for energy data collection
+- Based on detected firmware version enable the supported features
+- API details about supported data is combined into api.py
+- Added tests
```
[the remark this answers is not in the document]
Comment (bouwew) [reply CHANGELOG.md]: Fixed by adding a blank line

**Review by ArnoutD — APPROVED** [review]
Good for main, this branch can run for as long as you want... current main requires daily restarts

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,18 +1,42 @@
...
+  - Correct setting of accept_join_request which enables the auto-joining temporarily
+  - Change NodeAddRequest to not expect a response, the NodeRejoinResponse is often delayed past the NODE_TIMEOUT
+  - Use the already present response-subscription to capture the NodeJoinAvailableResponse and initialize to joining of a Node
+- Improve async task-handling, this should stress the CPU less (@bouwew)
+- Limit cache-size to 24hrs instead of to 1 week (@bouwew, @ArnoutD)
+- Implement support for devices with production enabled (@bouwew)
+- Collect Stick NodeInfo for use in HA (@bouwew)
+- Several bugfixes (@ArnoutD)
+- Github improvements/fixes (@CoMPaTech)
+- Async fixes (@ArnoutD - #141)
```
[the remark this answers is not in the document]
Comment (bouwew) [reply CHANGELOG.md]: Thanks, implemented.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,18 +1,42 @@
...
+
+- Make auto-joining work: (@bouwew)
+  - Correct setting of accept_join_request which enables the auto-joining temporarily
+  - Change NodeAddRequest to not expect a response, the NodeRejoinResponse is often delayed past the NODE_TIMEOUT
+  - Use the already present response-subscription to capture the NodeJoinAvailableResponse and initialize to joining of a Node
+- Improve async task-handling, this should stress the CPU less (@bouwew)
+- Limit cache-size to 24hrs instead of to 1 week (@bouwew, @ArnoutD)
+- Implement support for devices with production enabled (@bouwew)
+- Collect Stick NodeInfo for use in HA (@bouwew)
+- Several bugfixes (@ArnoutD)
```
[the remark this answers is not in the document]
Comment (bouwew) [reply CHANGELOG.md:13]: Implemented.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,18 +1,42 @@
...
+- Github improvements/fixes (@CoMPaTech)
+- Async fixes (@ArnoutD - #141)
+
+## v0.40.0 (a22)
+
+- Correcting messageflow to HA
+
+## v0.40.0 (a4)
+
+Full rewrite of library into async version (@brefra).
```
[the remark this answers is not in the document]
Comment (bouwew) [reply CHANGELOG.md:23]: Yes, @brefra did an amazing job to lay the basis for the async-version :+1:
